src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs

```diff
@@ -35,8 +35,50 @@ public static class TaskTimeoutExtensions
             }
             else
             {
-                throw new TimeoutException();
+                throw new TimeoutException($"Task timed out after {millisecondsTimeout}");
             }
         }
+
+        public static async Task WhenAllOrAnyFailed(this Task[] tasks, int millisecondsTimeout)
+        {
+            var cts = new CancellationTokenSource();
+            Task task = tasks.WhenAllOrAnyFailed();
+            if (task == await Task.WhenAny(task, Task.Delay(millisecondsTimeout, cts.Token)).ConfigureAwait(false))
+            {
+                cts.Cancel();
+                await task.ConfigureAwait(false);
+            }
+            else
+            {
+                throw new TimeoutException($"{nameof(WhenAllOrAnyFailed)} timed out after {millisecondsTimeout}");
+            }
+        }
+
+        public static Task WhenAllOrAnyFailed(this Task[] tasks)
+        {
+            int remaining = tasks.Length;
+            var tcs = new TaskCompletionSource<bool>();
+            foreach (Task t in tasks)
+            {
+                t.ContinueWith(a =>
+                {
+                    if (a.IsFaulted)
+                    {
+                        tcs.TrySetException(a.Exception.InnerExceptions);
+                        Interlocked.Decrement(ref remaining);
+                    }
+                    else if (a.IsCanceled)
+                    {
+                        tcs.TrySetCanceled();
+                        Interlocked.Decrement(ref remaining);
+                    }
+                    else if (Interlocked.Decrement(ref remaining) == 0)
+                    {
+                        tcs.TrySetResult(true);
+                    }
+                }, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
+            }
+            return tcs.Task;
+        }
```

Comment:
Has this not been rebased on the other changes yet?  I'll hold off on re-reviewing.

Reply:
Should be rebased now, if my git fu didn't fail me.